### ISSUE TITLE
Dockerfile could be optimized for better incremental builds

### ISSUE THREAD

**Issue by marcelo-gonzalez** [near/nearcore#8480]
Right now we have a `COPY . .` before we call `make` and build `neard` in the Dockerfile. This means that touching any file at all (say some random pytest file) will cause a full rebuild of everything, which is not great. It should be possible to refactor the Dockerfile to avoid this, with something like:

```
COPY ./Cargo.toml ./Cargo.lock ./
# build the deps here so we dont have to do this again
RUN make ...
# maybe this part could be a make target, like make clean-near-packages or something
# basically we want to remove everything but the deps from target/
RUN cargo clean --release --package {all the packages we care about}
# remove the auto generated src/ dir
RUN rm -r src/

# now copy everything and run make
COPY . .
RUN make ...
```
